### ISSUE TITLE
Explore user profiles

### ISSUE THREAD

**Issue by miguelsolorio** [microsoft/vscode#116740]
## Overview
This issue aims to explore what adding "profiles" could look like in the product. These profiles can be used in education, presentations, or can even be language specific. The explorations should cover being able to setup and switch between profiles. Ideally these profiles should be synced with [Settings Sync](https://code.visualstudio.com/docs/editor/settings-sync) (if used) so you can use those profiles anywhere.

One idea for settings is to introduce a new settings object like:

```jsonc
{
    "profiles": {
        "demos": {
            "files.autoSave": "onFocusChange",
            "editor.fontSize": 18,
            "editor.tabSize": 2,
            "editor.fontWeight": "600",
            "screencastMode.enabled": true,
        }
    },
    
    "editor.fontFamily": "Cascadia Code",
    "editor.renderWhitespace": "all",
}
```

The "Demos" profile inherits the current settings a user has and adds additional settings on top of the current ones. Also, when profiles are enabled we will create a "Default" profile where the user can switch back to their default mode.

## Related issues
- Screencast Mode only settings #105226
- Sync mutliple settings profiles #92393
- Multiple environment profiles (either per-session or per-window) #57548
- User profile #53453
- Enable/disable extensions from config file #40239
- Local Workspace settings #40233
- Support "perspectives" like approach in VSCode #95182

**Comment by misolori** [comment]
Here's the initial exploration of what profiles could look like, starting with a basic example of setting up a "Demos" profile. In this profile, the user can choose to either start with an empty profile (no settings/extensions) or keep their existing data. Settings could look something like this:

```jsonc
{
    "[profiles]": {
        "[Demos]": {
            "profiles.data": "keep",
            "files.autoSave": "onFocusChange",
            "editor.fontSize": 18,
            "editor.tabSize": 2,
            "editor.fontWeight": "600",
            "screencastMode.enabled": true,
        }
    },
    
    "editor.fontFamily": "Cascadia Code",
    "editor.renderWhitespace": "all",
}
```

The "Demos" profile inherits the current settings a user has and adds additional settings on top of the current ones.

## Demo
This demo shows the entry point from the gear menu, but it could also be enabled via the command palette. I didn't show what extension management looks like (yet) because it requires a bit more thinking.

![Kapture 2021-02-17 at 22 11 18](https://user-images.githubusercontent.com/35271042/108313565-608a9180-716d-11eb-8316-72a900c45ff0.gif)

**Comment by misolori** [comment]
## Concept B

This concepts attempts to re-use the accounts menu in the activity bar and also creates a default profile and reduces the effort involved with creating/switching accounts.

![Kapture 2021-02-18 at 09 00 38](https://user-images.githubusercontent.com/35271042/108393403-77120680-71c8-11eb-86e6-9f2de89fd46a.gif)

**Comment by carlocardella** [comment]
> One idea for settings is to introduce a new settings object like:
> 
> ```js
> {
>     "profiles": {
>         "demos": {
>             "files.autoSave": "onFocusChange",
>             "editor.fontSize": 18,
>             "editor.tabSize": 2,
>             "editor.fontWeight": "600",
>             "screencastMode.enabled": true,
>         }
>     },
>     
>     "editor.fontFamily": "Cascadia Code",
>     "editor.renderWhitespace": "all",
> }
> ```
> 
> The "Demos" profile inherits the current settings a user has and adds additional settings on top of the current ones. Also, when profiles are enabled we will create a "Default" profile where the user can switch back to their default mode.

So with this approach if I update a setting in the default profile (say `editor.fontFamily`) but I want the Demos profile to use the default font instead, it should be overridden with its default value here?

That would also mean that every change made in the default profile is automatically inherited by all other profiles? If the change was intended only for the default profile, all other profiles should be modified to revert that change?

I realize it is probably unlikely someone will have a high number of profiles (but who knows, really?); one way to avoid the _revert_ problem above would be to not inherit from the default profile, maybe only use it as a starting point when a new profile is first created. Of course that also means if I change a setting in the default profile that I _do_ want inherited, I'm out of luck... Maybe this _inherit from default profile_ could be a setting as well?

**Comment by jjspace** [comment]
> Maybe this inherit from default profile could be a setting as well

What if it was set up in a similar way to how `.eslintrc` can "extend" other configurations? 

For example, when you create a new profile it starts with an `"extends-profile": "default"` setting which would mean it always inherits from the default, even if default changes. If you no longer want this behavior just remove this extends setting and this profile would go back to App defaults for everything that's not set for that profile?

This would also allow profiles to extend each other (and maybe even allow people to share them!). To go with OPs example, maybe I have a `teaching` profile with bigger font/zoom and some parts hidden. But now I start teaching another class in a different language and I want a different set of extensions or maybe a different theme that's better for this language. I could just create a new profile that _extends_ my base `teaching` profile with just the new settings.

I guess this would still leave the question of "if I change a setting of the 'child' profile, which level of the config do I want it on, this one or the parent?" Maybe it's still best to have no inheritance at all and all new profiles should start with App defaults or duplicate the default at creation.

**Comment by jjspace** [comment]
Also not directly related to current conversation but the term "Profiles" threw me off at first. I was expecting something along the lines of User profiles like multiple Chrome accounts. Not different configuration sets. I think profiles works but it _may_ be worth thinking of alternatives?

**Comment by misolori** [comment]
Yea I did originally play around with the "inheritance" concept, mostly focused on disabling it in a certain profile. The idea of mixing profile sounds interesting but would love to see a use case for it first because it could add some unnecessary complications, like you just described.

In terms of the term "Profiles", this one is tricky because we do also have user accounts where you login for certain extensions (like GitHub or Azure), so we want to make sure those are separated and not mixed. But thanks for this feedback, we'll gauge to see if we get similar feedback.

**Comment by yume-chan** [comment]
I prefer concept B. Looking together with #115641, maybe there is no such strong need for the account manager. I personally never touched it after setting up all my accounts.

However, from these early concepts, I did get how it will resolve all related issues.

* Can I use different profiles for different Code instance? Will it switch profile automatically when opening a workspace or enter/exit screencast mode?
* I care more about #40233, what's the priority of profile settings? Will it be like user settings < profile user settings < workspace settings < workspace profile settings?

**Comment by gjsjohnmurray** [comment]
Another related issue is #95182, which proposes the term "perspectives".

**Comment by LucaGabi** [comment]
> Here's the initial exploration of what profiles could look like, starting with a basic example of setting up a "Demos" profile. In this profile, the user can choose to either start with an empty profile (no settings/extensions) or keep their existing data.
> 
> ## Concept A
> This concept shows the entry point from the gear menu, but it could also be enabled via the command palette. I didn't show what extension management looks like (yet) because it requires a bit more thinking.
> 
> ![Kapture 2021-02-17 at 22 11 18](https://user-images.githubusercontent.com/35271042/108313565-608a9180-716d-11eb-8316-72a900c45ff0.gif)

This looks very good and easy to use also.
B is hard to navigate in my opinion.

**Comment by NowanIlfideme** [comment]
Hi, would this "user profile" concept work for multiple accounts in vscode itself? For example, I use vscode for personal projects (personal email & GitHub account, this one), but also for work (my employer's msft-based email & work GitHub account), potentially also integrating with a customer's system (e.g. _their_ Microsoft account). However, I want my settings to sync via my personal email (for example), or at least to start both from the same config.

**Comment by RobJacobson** [comment]
Great ideas!  I'm looking for a simple tool that allows me to switch profiles/extensions between my HTML/CSS/JS stack, my Python stack, and my Ruby stack.  There's no need for one set of extensions to coexist with the other sets of extensions, and it's really tedious to enable/disable extensions by hand.  

For example, when coding Ruby, I don't need ESLint to be enabled as an extension.  I would prefer it ESLint just "didn't exist" in that VS Code workspace (not merely be disabled).

There is already an extension that does much of this, [Profile Switcher](https://marketplace.visualstudio.com/items?itemName=aaronpowell.vscode-profile-switcher).  Unfortunately, it's still in version 0.3.3 and has been since 2019.  It appears to be abandonware.  It might be buggy, or I might be buggy ([PEBMAC](https://en.wikipedia.org/wiki/User_error)), but it seems to have "forgotten" a large number of my favorite extensions.  Now I need to do a "hard reset" of VS Code and reinstall them.

https://www.aaron-powell.com/posts/2019-06-28-vscode-extension-for-managing-profiles/

There are articles/blog posts at the links below about how to switch user profiles from the command line.  However, this also seems rather tedious.
https://dev.to/sulmanweb/vs-code-pro-tip-code-profiles-multi-environment-development-hg2
https://www.reddit.com/r/vscode/comments/c2h4a7/different_profiles/
https://dev.to/jsjoeio/how-to-create-code-profiles-in-vscode-3ofo
https://www.freecodecamp.org/news/how-to-create-code-profiles-in-vscode/

Could the VS Code team please consider making this an available option?  It could be as simple as taking the existing [Profile Switcher](https://github.com/aaronpowell/vscode-profile-switcher/blob/main/License.md) code (MIT licensed) and cleaning it up.

**Comment by RobJacobson** [comment]
There's also another extension, "User Profiles," that seems to have been released just a few days ago.
https://github.com/cyberbiont/vscode-profiles

**Comment by U-C-S** [comment]
Any progress on this one?

**Comment by Chaphasilor** [comment]
I simply wanted to add another "User Story" for this feature, as well as a more detailed implementation strategy

Similar to what @RobJacobson said, I'm looking for a way to first and foremost configure profile-specific extension sets.  
**My use-case:**

Working with "writing" tools like LaTeX or markdown is often an involved process, with numerous heavy-weight extensions being required. However, 95% of the time I'm *not* writing and therefore not using these extensions. Disabling and re-enabling these extensions manually would be a solution, but a very tedious, lengthy and error-prone one.  
Switching profiles, including active extensions and settings (e.g. `Word Wrap Column` in this case), on a **per-workspace** basis, would be the perfect solution.  
Using workspace-specific settings would also be possible, but in this case I'd need to manually "duplicate" these settings if I start a new writing project that is unrelated to the others, and I dislike having multiple folders added to a single workspace :)

**What I would suggest** is giving us the profile functionality, **tied to workspaces**, so that we can select a specific profile for each workspace. A "default" profile could be used well, by default, and changing the default profile would cause all workspaces, which use the default profile, to change as well.

This should cover most use cases with a very simple concept that is completely compatible with Concept A (status bar items, like the current git branch or Workspace Trust, are also tied to the workspace).

**Comment by evandcombs** [comment]
For the name of this feature may I suggest 'Environments'.

**Comment by gjsjohnmurray** [comment]
Linking to https://code.visualstudio.com/updates/v1_67#_settings-profile

**Comment by picwellwisher12pk** [comment]
I like to contribute my idea. 
My idea was to have this as part of a project. Maybe part of `extensions.json` in .vscode directory.


![image](https://user-images.githubusercontent.com/1508646/168021751-446328dc-26a7-4222-985b-adfd14dca86d.png)
**extensions.json inside my node.js based project** 

I believe this is simpler from a user's point of view

**Comment by hoa00** [comment]
> I like to contribute my idea. My idea was to have this as part of a project. Maybe part of `extensions.json` in .vscode directory.
> 
> ![image](https://user-images.githubusercontent.com/1508646/168021751-446328dc-26a7-4222-985b-adfd14dca86d.png) **extensions.json inside my node.js based project**
> 
> I believe this is simpler from a user's point of view

This idea is awesome

**Comment by Sylk** [comment]
Would love to see this as a feature, it's difficult to have so many different stacks in a single editor. The intellisense starts playing goofy if I can't sandbox it down.

**Comment by gjsjohnmurray** [comment]
After switching to a different profile it can take a few seconds before all the changes have been applied. It would be good to have a progress indicator. Perhaps a spinner in the status bar item. Or a progress notification, which could perhaps also inform us a seach part of the switchover is being performed.

**Comment by sandy081** [comment]
Hi all, the latest [Insider](https://code.visualstudio.com/insiders/) release has support for settings profiles. It would be great if you can try it out and provide us feedback through issues or comments in this issue. You can enable this feature in insiders using the setting `workbench.experimental.settingsProfiles.enabled`.

**Comment by sandy081** [comment]
I do not think so, Accounts are handled across profiles.

CC @TylerLeonhardt

**Comment by TylerLeonhardt** [comment]
@iRoachie you'll want to follow along with this issue: https://github.com/microsoft/vscode/issues/127967#issuecomment-1111420554

We'll need that before we can do a "different github accounts per profile"

**Comment by xiaoxi-david** [comment]
It would be very interesting if VS Code came with preconfigured profile settings including official and most popular extensions for the most popular programming languages.

**Comment by sandy081** [comment]
@Kevin-Delnoije Thanks for trying out and providing feedback.

> i could not find a way however to choose profiles from the command line e.g. code --profile profile.code-profile so that i can alias different profiles in termina

CLI support for profiles is in progress.

> Another thing i would love to see is a global config for settings like theme and keybindings which are mostly independent of different projects. maybe giving multiple profiles and merging could be a solution? code --profile global.code-profile project.code-profile

I see your point and heirarchies in profiles will make them complicated to use and understand. So we will collect more feedback here to understand more about this.

**Comment by femto-code** [comment]
I am interested in seeing the user profiles evolving. From a developers point: how is switching between different profiles accomplished **in terms of extensions**?
Assuming you have different sets of extensions there seem to be _two options_ for me:
1. Are extensions uninstalled, so deleted from disk? This would mean new download on next switch.
2. Or maybe cached? Maybe just moved away from `extensions` directory? This way an extensions would be moved in place if selected in user profile.

**Comment by femto-code** [comment]
I agree with @Kevin-Delnoije in the matter of a few additions necessary in order to make the settings profiles adoptable to one's own use case:

1. Allow binding a settings profile to a workspace/directory (preferably via workspace/folder `settings.json`, maybe include the `.code-profile there`)
2. Provide a way of having a global configuration i.e. extensions or settings/keybindings that would be independent from profiles (the application-level settings also go here).

**Comment by sandy081** [comment]
Extensions are not uninstalled or installed (downloaded) during profile switching. It is just loading the set of extensions listed from the profile that is active.

**Comment by femto-code** [comment]
@sandy081 Use case would be simple: I have multiple workspaces and don't want to have individual profile settings that I have to maintain separately.
I think it would be great to give the opportunity so select what to include in the `.code-profile`. E.g. if settings are not included then the global settings file in user data dir is used. Another option would be to have an auto merge option.
At least global settings would be a must-have for me.

**Comment by scnwwu** [comment]
Say I've created multiple profiles. Now I found a new theme (or new extension) which is awesome and I'd like to use it globally. In current Insiders version, I have to manually apply in each profile right?
One idea would be adding another tab say "Profile" in the Settings page beside User and Workspace. Change the settings in the User tab will affect all profiles for the user. And there can be "Enable (Profile)" beside "Enable (Workspace)" in the extension menu.
Another idea would be to have a "Base" profile. The extensions enabled in the Base profile work across all profiles. Any settings that's not a default value in Base profile work across all profiles, something like that.
I think it's a common case that user want some settings applied globally and some per profile.

**Comment by alefragnani** [comment]
> Extensions are not uninstalled or installed (downloaded) during profile switching. It is just loading the set of extensions listed from the profile that is active.

Is there any chance to expose this _loading set of extensions_ feature/API to `vscode.openFolder` command? 

Today, this command only supports an `Uri` describing the folder/workspace path to be opened. It would be interesting to allow adding the set of extensions to be used as well, or at least the `profile`, just like you are exposing in #151301.

Thank you

**Comment by sandy081** [comment]
@femto-code @scnwwu Thanks for the feedback. Extracted the discussion into a separate issue - https://github.com/microsoft/vscode/issues/156144. Please continue discussion there.

**Comment by sandy081** [comment]
@alefragnani Please file a separate issue for your requirement

**Comment by alefragnani** [comment]
Filed #156173 as requested. Thank you!

**Comment by gusbemacbe** [comment]
@alefragnani and @sandy081 

I want to create a new current profile or export a profile **with workspace settings** instead of user settings. Why? Because the user settings are **already global** in all projects. I would rather not include Java settings in the global user profiles, but in a Java workspace settings. 

I already tried to explain you in te past regarding #90649, but my autism limits my communication.

**Comment by gusbemacbe** [comment]
I do not recommend Microsoft's VSCode's native Settings Profiles!

I used to enabled more or less than 120 extensions for basic Markdown projects on a workspace. I exported the settings profile, then I imported into… **It got a disaster!** It enabled almost all the extensions for a basic Markdown profile!

![image](https://user-images.githubusercontent.com/2257024/182034057-4235e039-66ac-4d45-8d13-0aaf81fd1dce.png)

I really miss @aaronpowell's depreciated extension [VSCode Profile Switcher](https://github.com/aaronpowell/vscode-profile-switcher) that was totally functional.

**Comment by gusbemacbe** [comment]
I made another test, importing a basic Kotlin profile. It really enabled all extensions. Not only it, when you import the settings profile, …:

- it will install all the extensions even if they are already installed in the `~/.vscode-insiders/extensions`!
- you'll be asked to change more than seven languages, to insert CodeTime and WakaTime tokens, etc.

![image](https://user-images.githubusercontent.com/2257024/182036088-b2306462-2509-4fbc-8648-dd52ce536237.png)


Totally disastrous.

**Comment by rockboynton** [comment]
suggested setting for enabling this feature is not recognized for me 
![image](https://user-images.githubusercontent.com/32227280/182934135-d2ffb7c7-38a0-4a0b-86ff-e326117dcfe3.png)

**Comment by carlocardella** [comment]
This feature is still in preview in [Insiders](https://code.visualstudio.com/insiders/) only, please check the VSCode version you are using

**Comment by rockboynton** [comment]
> @rockboynton this feature is still in preview in [Insiders](https://code.visualstudio.com/insiders/) only, please check the VSCode version you are using

got it, thanks!

it's a bit confusing because some of the experimental features just need to be enabled and some need insiders, and that's not always specified

**Comment by markovicdenis** [comment]
Would it be possible to switch profiles automatically based on system theme mode (dark/light)? Someone probably needs to make an extension for that, since it would be out of the scope for this feature?

**Comment by sandy081** [comment]
> Would it be possible to switch profiles automatically based on system theme mode (dark/light)? Someone probably needs to make an extension for that, since it would be out of the scope for this feature?

@markovicdenis Thanks for the idea. But, there are no plans to support that in core.

**Comment by kamenminkov** [comment]
The setting is still not recognised in the latest Insiders 1.71 version as of today; was it disabled at some point?

**Comment by carlocardella** [comment]
Interesting, it may have been hidden by mistake in a recent Insiders release, I still see the menu commands (maybe because I had enabled the setting before) but it is gone from the configuration: 

![image](https://user-images.githubusercontent.com/5784415/187220119-f79f5436-d185-4b4a-af13-1b95e140fb95.png)

![image](https://user-images.githubusercontent.com/5784415/187220714-1ace6512-ee90-4269-b761-a9a123c931ef.png)

**Comment by carlocardella** [comment]
It looks like in Insiders the setting is no longer needed? This is a brand new setup on a brand new VM (to avoid settings cached somewhere) and Settings Profiles is available out of the box: 

![image](https://user-images.githubusercontent.com/5784415/187227700-b689f17b-7e04-484a-b3b8-7d6e7d2c1db2.png)

**Comment by sandy081** [comment]
We enabled this feature by default in insiders and from next release this is also enabled in stable but behind the flag.

**Comment by kamenminkov** [comment]
Thanks for the explanations - it makes sense now.

Started playing around with those - and what comes to mind immediately is that it will be super useful to have some inheritance/overriding pattern (e.g. in a similar fashion to what we now have - how a workspace setting can override a user setting). The way I see it now is that you have to maintain two sets of settings with no way of providing a common base to them to inherit from - or perhaps I'm missing something else? No doubt the current state would come in handy for people who need to maintain different settings and extensions for different tech stacks, but IMO there needs to be a way to have a common base too; if I'm jumping between technologies, I personally would like to preserve for example font and colour settings, some of my extensions, most of my keybindings and things like that - and I believe it would be the same for other people.

**Comment by daviddossett** [comment]
> there needs to be a way to have a common base too; if I'm jumping between technologies, I personally would like to preserve for example font and colour settings, some of my extensions, most of my keybindings and things like that

We're certainly thinking about this. Stay tuned 👍

**Comment by jcrben** [comment]
When I export it looks like this - does it really need all these `\\\`? 
<img width="971" alt="image" src="https://user-images.githubusercontent.com/5614134/188282934-bbb88957-d431-4e3f-a17f-93a27e025535.png">


I will probably be looking to commit and edit these code-profiles

**Comment by jmevel** [comment]
Hello guys,

First of all thanks for all the work you've done to make this new feature, I think many of us were waiting for it.
Now there's something I'm a bit confused, I'm not sure that's even possible with this first version of the profiles feature.

Is it possible to enable/disable extensions based on the profile we use?
When I disable an extension I don't see any change  in the user settings neither in the profile settings.
Searching on internet it seems [it's not yet possible](https://github.com/microsoft/vscode/issues/40239) so does it mean the profile feature doesn't have this ability yet? Do you have this feature tracked in your backlog already?

What I'd personally love is to be able to enable/disable extensions simply based on the selected profile, not just based on the current opened workspace. The simple use case of it is to easily switch settings between work and home as my needs are very different in these two environments (however I still want to sync all my settings so I can easily retrive my work/home profiles from any computer).

I hope this feature could be added to your backlog and tracked for future release, I think I'm not the only one waiting for it 🙂

Thanks for your answers!

**Comment by ZaLiTHkA** [comment]
> Is it possible to enable/disable extensions based on the profile we use?

hey @jmevel, have you played with these "Profiles" yourself yet, or are you checking details before you do..?

I've just tested on my side by creating a new empty profile, and VSCode switched to what looks like a fresh installation with none of my explicitly installed third-party extensions.

switching back to "Default" re-enabled everything again as I had it. this sounds exactly the same as what you're asking about.

---

..and while I'm here, just some general feedback: I've been following this thread for ages, but I haven't actually tried this feature until now.. going to spend some time now setting things up.

I regularly switch between project types:

* IDE-agnostic Maven projects, which the ["Extension Pack for Java"](https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-pack) helps with.
* Netbeans-specific Ant projects, which ["Langage Server for Java by Apache NetBeans"](https://marketplace.visualstudio.com/items?itemName=ASF.apache-netbeans-java) helps a lot with.
* Dart/Flutter projects, some of which also use Java, specifically for Android builds.
* "static web" projects (React, Angular, etc.), mostly coordinated using NodeJS or Deno.

suffice to say, I am very much looking forward to having VSCode **not** initialise my collection of Redhat Java extensions whenever I open my Ionic Cordova projects... 😁

**Comment by jmevel** [comment]
> > Is it possible to enable/disable extensions based on the profile we use?
> 
> hey @jmevel, have you played with these "Profiles" yourself yet, or are you checking details before you do..?
> 
> I've just tested on my side by creating a new empty profile, and VSCode switched to what looks like a fresh installation with none of my explicitly installed third-party extensions.
> 
> switching back to "Default" re-enabled everything again as I had it. this sounds exactly the same as what you're asking about.

You are right sir, I'm sorry for my useless comment. 

I guess I'm too used to configuring eveything in the `settings.json` file but also I think I was fooled by what's written when selecting any extension `This extension is enabled globally` or `This extension is disabled globally by the user`. The `globaly` is a bit misleading now in my opinion...

Anyway, thanks for your answer, this is a wonderful news 😇

**Comment by jmevel** [comment]
However one thing I just noticed and absolutely need to be added in my opinion is the synchronization our profiles!
I was surprised yesterday evening not to find my profiles on my personal laptop and indeed it seems not possible to configure it for syncing. (The `workbench.experimental.settingsProfiles.enabled` setting was also missing even though it's in the default `settings.json` file)

![image](https://user-images.githubusercontent.com/2572775/188808521-2b329836-7d2a-45a6-8c30-f7e85f344490.png)

Is it because this feature is still in `experimental` state for the moment?

**Comment by sandy081** [comment]
We are planning to support syncing profiles in current (September 2022) milestone - https://github.com/microsoft/vscode/issues/154243

**Comment by PeterWone** [comment]
A long time coming but well thought through and neatly integrated. With the imminent profile syncing and support for extension installation/enablement, you have produced a capability (usable roaming profiles) that Windows has promised since the nineties and never quite delivered. This is no small achievement.

**Comment by Oh-Turry** [comment]
Is there any chance of this to work in WSL?

**Comment by Chaphasilor** [comment]
I noticed that the `window.autoDetectColorScheme` setting doesn't work well with user profiles. When my laptop switches from dark to light theme, previously all instances would be switched to the light theme (as they all shared the same settings).  
Now with the settings profile, only the currently "active" profile (i.e. profile used by the first VS Code instance/window I focus after the OS theme change) is switched to light mode, all other profiles, whether open or closed, continue to use the dark theme until manually switching theme.  
Same goes for the other way around, light theme -> dark theme.

I'm guessing this is due to the profiles not yet having a shared "base profile" that they all dynamically inherit from. I think this was discussed previously, as it would make a lot of sense. Aside from the automatic theme switching, things like common keyboard shortcuts should be the same in all profiles, and it would be nice to have an option to selectively enforce this somehow :)

**Comment by equiman** [comment]
I'm trying to use the new settings profile. I'm actually on the "Default" profile.

And I'm trying to create a new one from blank with "Create Empty Settings Profile..." but when I open the settings editor the settings are the same (but grayed)

Default (Left) vs New Profile (right)
![image](https://user-images.githubusercontent.com/933393/192187113-64d01c59-8c0e-4569-aec8-6dbe429d022d.png)

I see that the extension is the only thing that is different between the profiles, but not the settings file.

- If it's empty why is showing the previous configurations on the new one?
- Why when I change values on the JSON configuration in the new profile and switch to default those changes are applied to the also? They are not independent?

**Comment by gjsjohnmurray** [comment]
@equiman after you ran the command to create the empty profile did you switch to it? Look on the gearwheel menu at the foot of the Activity Bar. There's probably a Command Palette entry for doing that too. When your profile is active you should see a clickable 2-letter 'pill' below the gearwheel showing the first 2 letters of your profile's name. Now just change user settings normally, or enable / disable extensions.

**Comment by sandy081** [comment]
Thanks for providing some info @gjsjohnmurray, but @equiman is on stable and the pointers you recommended are in insiders.

@equiman Thanks for trying profiles. You are right that every profile has its own settings. May I know how you are opening settings (file) when you are in a window that is not using default profile? If you are using settings UI, you shall see two options to open profile specific settings and application specific user settings on the top right corner

![image](https://user-images.githubusercontent.com/10746682/192219058-7dfa18e0-6ea1-4071-8dcb-3e825651fbe1.png)

You can open profile specific settings file by using the first action from above picture. Every profile (except default) uses default settings.json file for configuring application scoped settings like application update configuration. Hence when you opened the default settings.json in a window with custom profile, all those settings which are not application specific are dimmed mentioning that they are not applicable in this window. Any changes in the default settings.json file will effect the default profile. To customise profile specific settings, you have to open profile specific settings file as shown above. I agree this is bit confusing but this is mitigated by Settings UI showing them as expected by the User. 

Please let me know if it clear. 
Thanks

**Comment by jmevel** [comment]
@equiman 
As @sandy081 already pointed out, you must open the settings file of your current profile after you've switched.

![VS Code](https://user-images.githubusercontent.com/2572775/192228822-4f039362-5035-44cf-89a7-c9011a661a5b.PNG)
Notice on the right one it has a generated ID in its path because that's a user profile and on the left it doesn't have it because that's the default settings.

I advise you to write a comment on top of each file so you can quickly make sure which settings file you're currently viewing/editing.

**Comment by equiman** [comment]
Thanks @sandy081 ok, perfect now is clear, how to open the profile settings.
![image](https://user-images.githubusercontent.com/933393/192284905-9a005af5-78d3-4350-8253-7c88782ab354.png)

Now I can see and change the file related to this profile
![image](https://user-images.githubusercontent.com/933393/192285651-1c7a76c8-2ea8-4693-a638-5deece11f51f.png)

@jmevel thanks for the comment tip

**Comment by isidorn** [comment]
My name is Isidor and I am helping @sandy081 with the profile work.
Right now we are interested in feedback regarding sharing profiles. So if you have 30 mins to chat and are:
* An Educator looking to use profiles in their class room
* Engineer planning to share profiles in their 10+ person team

Please drop me an email at inikolic@microsoft.com so we get together and quickly talk.
Thanks 🙏

**Comment by ZaLiTHkA** [comment]
> My name is Isidor and I am helping @sandy081 with the profile work. Right now we are interested in feedback regarding sharing profiles.

just for clarification, is there any plan to work on the concept of "profile inheritence"? for example, profile "My Cool Profile" that inherits the configuration defined in "My Base Profile", which might in turn inherit from "Default"..

I spent some time configuring a few profiles the other evening, starting from a completely blank slate: I first configured the built-in "Default" profile, installing a few common extensions and setting some personal preferences. then while on "Default", I created another using the "Create from Current Settings Profile..." option. I then switched back to "Default" and did this again for another profile, adding different extensions and setting more custom preferences.

tl;dr: the amount of overlap between my profiles is incredibly high, and I'm now concerned that I've created a maintenance monster in my IDE. 🤔

**Comment by ZaLiTHkA** [comment]
thanks @isidorn, I can fully appreciate wanting to avoid profiles inheriting from other custom profiles.. but optionally setting one custom profile to inherit from the internal "Default" profile would be absolutely epic.

**Comment by mrwensveen** [comment]
It would be cool if we could source in another json file, be it another profile or just another json file. That would basically give you all the inheritance functionality you'd need.

**Comment by equiman** [comment]
Is there a way o flag to set on vscode project settings to set the profile to use on this project?

**Comment by AverTry** [comment]
Please have default extensions reflect changes to created from.

If I have a basic set that I would need in all profiles, then I don't have to make changes across multiple profiles, manually.
Or a way to drag n drop between profiles.
As I just had to make changes across 5 profiles and adjust key-shortcuts in each.

Thanks for this amazing feature, and it's possibilities, 
because I always mess up the settings sync, and makes me not use it.

**Comment by moseleyi** [comment]
Will we be able to assign a profile to language/folder/workspace?

**Comment by kamenminkov** [comment]
> Will we be able to assign a profile to language/folder/workspace?

The way I see it you could already assign settings to a language even before profiles (e.g. the `"[<language>]"` syntax in `settings.json`) whereas the idea of profiles as I get it is for them to be easily switchable. I guess you can expand on that idea by using language-specific settings inside a profile's `settings.json`. As for folders and workspaces, some previous comments mention that this is already working like that.

**Comment by carlocardella** [comment]
> We are planning to support syncing profiles in current (September 2022) milestone - #154243

How should profile sync work? I assume profiles created on one machine should be sync'ed to other logged in with the same account (GitHub, in my case)? Sync is working for all settings except profiles (I am using the latest Insiders).

```
Version: 1.73.0-insider (user setup)
Commit: fad3a77833b9249158dfd88477114a06435e46a2
Date: 2022-10-07T05:18:17.049Z
Electron: 19.0.17
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Windows_NT x64 10.0.22621
Sandboxed: Yes
```

No errors or any reference to "profile" in `Log (Setting Sync)` Output panel or in `Sync Activity`

![image](https://user-images.githubusercontent.com/5784415/194612776-e19cbffa-6a8f-4f01-a1ba-82c67274267c.png)

![image](https://user-images.githubusercontent.com/5784415/194613387-9abb777f-97de-4702-9596-631ad86aecba.png)

**Comment by Thesoreon** [comment]
Hi 👋, it seems like the profile badge in the Activity Bar has a misaligned short name when the Activity bar is right positioned.

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/32848654/194770722-3b35429d-8c2c-4f83-9212-4d4311c7364c.png">

**Comment by gjsjohnmurray** [comment]
> Hi 👋, it seems like the profile badge in the Activity Bar has a misaligned short name when the Activity bar is right positioned.
> 

Already reported and fixed in Insiders.

**Comment by sandy081** [comment]
@carlocardella Can you please share the settings sync logs (from Output panel) from both machines?

**Comment by carlocardella** [comment]
> @carlocardella Can you please share the settings sync logs (from Output panel) from both machines?

#### Source machine (where I have created the profile I want to sync):

```
[2022-10-11 16:54:52.688] [settingssync] [info] Using settings sync service https://vscode-sync.trafficmanager.net/
[2022-10-11 16:54:52.688] [settingssync] [info] Auto Sync is enabled.
[2022-10-11 16:54:52.688] [settingssync] [info] Auto Sync: Suspended until auth token is available.
[2022-10-11 16:55:02.488] [settingssync] [info] Auto Sync: Triggered by Interval
[2022-10-11 16:55:02.489] [settingssync] [info] Sync started.
[2022-10-11 16:55:02.900] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 16:55:02.928] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 16:55:02.958] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 16:55:02.972] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 16:55:09.820] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 16:55:10.925] [settingssync] [info] Extensions: No changes found during synchronizing extensions.
[2022-10-11 16:55:10.931] [settingssync] [info] Sync done. Took 8443ms
[2022-10-11 16:57:01.333] [settingssync] [info] Auto Sync: Triggered by Activity
[2022-10-11 16:57:01.333] [settingssync] [info] Sync started.
[2022-10-11 16:57:01.485] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 16:57:01.500] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 16:57:01.520] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 16:57:01.530] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 16:57:01.538] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 16:57:01.791] [settingssync] [info] Extensions: No changes found during synchronizing extensions.
[2022-10-11 16:57:01.796] [settingssync] [info] Sync done. Took 464ms
[2022-10-11 16:58:27.225] [settingssync] [info] Auto Sync: Triggered by Activity
[2022-10-11 16:58:27.225] [settingssync] [info] Sync started.
[2022-10-11 16:58:27.449] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 16:58:27.459] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 16:58:27.489] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 16:58:27.511] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 16:58:27.523] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 16:58:28.116] [settingssync] [info] Extensions: No changes found during synchronizing extensions.
[2022-10-11 16:58:28.118] [settingssync] [info] Sync done. Took 894ms
[2022-10-11 17:01:08.871] [settingssync] [info] Auto Sync: Triggered by Activity
[2022-10-11 17:01:08.871] [settingssync] [info] Sync started.
[2022-10-11 17:01:09.002] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 17:01:09.013] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 17:01:09.038] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 17:01:09.049] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 17:01:09.058] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 17:01:09.300] [settingssync] [info] Extensions: No changes found during synchronizing extensions.
[2022-10-11 17:01:09.303] [settingssync] [info] Sync done. Took 433ms
[2022-10-11 17:04:50.759] [settingssync] [info] Auto Sync: Triggered by Activity
```

In case it matters, there is no "profiles" file apparently sync'ed here:
![image](https://user-images.githubusercontent.com/5784415/195220185-d45257c5-ab53-442f-9a78-63f29cb40ba2.png)

#### Target machine (where I expect the profile to show up)

```
[2022-10-11 17:09:56.527] [settingssync] [info] Using settings sync service https://vscode-sync.trafficmanager.net/
[2022-10-11 17:09:56.527] [settingssync] [info] Auto Sync is enabled.
[2022-10-11 17:09:56.528] [settingssync] [info] Auto Sync: Suspended until auth token is available.
[2022-10-11 17:10:07.183] [settingssync] [info] Auto Sync: Triggered by Interval
[2022-10-11 17:10:07.184] [settingssync] [info] Sync started.
[2022-10-11 17:10:07.954] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 17:10:07.976] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 17:10:08.003] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 17:10:08.018] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 17:10:08.031] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 17:10:08.807] [settingssync] [info] Extensions: Skipped synchronizing extension because the extension is not found. ev2team.ev2-authoring-extension
[2022-10-11 17:10:08.817] [settingssync] [info] Extensions: Updated last synchronized extensions. Skipped: ["ev2team.ev2-authoring-extension","laurenttreguier.vscode-simple-icons","alanwalk.markdown-toc"].
[2022-10-11 17:10:08.822] [settingssync] [info] Sync done. Took 1640ms
[2022-10-11 17:10:14.792] [settingssync] [info] Auto Sync: Triggered by Activity
[2022-10-11 17:10:14.792] [settingssync] [info] Sync started.
[2022-10-11 17:10:15.277] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 17:10:15.290] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 17:10:15.324] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 17:10:15.347] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 17:10:15.578] [settingssync] [info] GlobalState: Updated remote ui state
[2022-10-11 17:10:15.587] [settingssync] [info] GlobalState: Updated last synchronized ui state
[2022-10-11 17:10:16.082] [settingssync] [info] Extensions: Updated remote extensions. Updated: ["ev2team.ev2-authoring-extension"].
[2022-10-11 17:10:16.090] [settingssync] [info] Extensions: Updated last synchronized extensions. Skipped: ["ev2team.ev2-authoring-extension","laurenttreguier.vscode-simple-icons","alanwalk.markdown-toc"].
[2022-10-11 17:10:16.095] [settingssync] [info] Sync done. Took 1305ms
[2022-10-11 17:15:16.101] [settingssync] [info] Auto Sync: Triggered by Interval
[2022-10-11 17:15:16.101] [settingssync] [info] Sync started.
[2022-10-11 17:15:16.403] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 17:15:16.415] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 17:15:16.443] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 17:15:16.462] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 17:15:16.477] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 17:15:16.931] [settingssync] [info] Extensions: No changes found during synchronizing extensions.
[2022-10-11 17:15:16.934] [settingssync] [info] Sync done. Took 833ms
```

Interestingly, the target machine has a `Settings Profiles` file
![image](https://user-images.githubusercontent.com/5784415/195221618-6b882b98-a475-4116-97ef-1c3bc13056ff.png)

This is its content:

```json
{"ref":"0","content":null}
```

Out of curiosity, I tried to create an empty profile with the same name on the target machine and force a sync, just to see what would happen if the profile already exists: still the sync does not seem to happen

Source machine:

```
[2022-10-11 17:18:58.886] [settingssync] [info] Sync started.
[2022-10-11 17:18:59.012] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 17:18:59.022] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 17:18:59.037] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 17:18:59.044] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 17:18:59.050] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 17:18:59.500] [settingssync] [info] Extensions: Updated remote extensions. Updated: ["ev2team.ev2-authoring-extension"].
[2022-10-11 17:18:59.527] [settingssync] [info] Extensions: Updated last synchronized extensions. Skipped: ["ev2team.ev2-authoring-extension","laurenttreguier.vscode-simple-icons","alanwalk.markdown-toc"].
[2022-10-11 17:18:59.534] [settingssync] [info] Sync done. Took 650ms
```

Target machine:

```
[2022-10-11 17:19:06.510] [settingssync] [info] Sync started.
[2022-10-11 17:19:06.739] [settingssync] [info] Settings: No changes found during synchronizing settings.
[2022-10-11 17:19:06.750] [settingssync] [info] Keybindings: No changes found during synchronizing keybindings.
[2022-10-11 17:19:06.784] [settingssync] [info] Snippets: No changes found during synchronizing snippets.
[2022-10-11 17:19:06.800] [settingssync] [info] Tasks: No changes found during synchronizing tasks.
[2022-10-11 17:19:06.811] [settingssync] [info] GlobalState: No changes found during synchronizing ui state.
[2022-10-11 17:19:07.426] [settingssync] [info] Extensions: Skipped synchronizing extension because the extension is not found. ev2team.ev2-authoring-extension
[2022-10-11 17:19:07.434] [settingssync] [info] Extensions: Updated last synchronized extensions. Skipped: ["ev2team.ev2-authoring-extension","laurenttreguier.vscode-simple-icons","alanwalk.markdown-toc"].
[2022-10-11 17:19:07.438] [settingssync] [info] Sync done. Took 932ms
```

**Comment by ZaLiTHkA** [comment]
> It seems both machines are using stable settings sync service - `Using settings sync service https://vscode-sync.trafficmanager.net/` and syncing profiles is not supported when using stable service.

so just to clarify, this means we **can** use "Settings Profiles" in VSCode Stable, but we **cannot** sync them across machines unless we use VSCode Insiders? looks like I'm going back to Insiders again.. 😎

**Comment by sandy081** [comment]
> so just to clarify, this means we can use "Settings Profiles" in VSCode Stable, but we cannot sync them across machines unless we use VSCode Insiders? looks like I'm going back to Insiders again.. 

Unfortunately Yes. This feature is still in preview and we are in the process of experimenting and getting feedback and stabilising it. We would like to have syncing enabled in insiders for a milestone before enabling it in stable, so that we can fix any strange issues.

> does VSCode Insiders store the "default" profile's extensions list differently now?

We do not create extensions.json if there are no profiles. This is being compatible with old behaviour when there are no profiles and all extensions are read from extensions installation folder. But we also have plans to make it consistent that default profile extensions are stored in extensions.json always. This needs some migration though.

**Comment by equiman** [comment]
> There is not such a flag, but you can open a workspace and associate the profile and this will be remembered. You can also associate a profile from cli. Eg: `code --profile <name> <folder>`

Well, while this tag was created. I'm using some aliases as you recommended. Thanks!

```bash
# VSCode
alias vc="code"
alias vco="vc ."
alias vd="code --profile \".Net\""
alias vdo="vd ."
alias vr="code --profile \"React\""
alias vro="vr ."
alias vn="code --profile \"Node\""
alias vno="vn ."
```

**Comment by hughlilly** [comment]
> I'm actually on the "Default" profile and I'm trying to create a new one with "Create Empty Settings Profile..."

How do I get to this option? I can’t find it… I only have “Import from…” and “Settings Profiles: Export (Default)…”.

**Comment by sandy081** [comment]
@hughlilly This is an experimental feature and can be enabled using the setting `workbench.experimental.settingsProfiles.enabled`

**Comment by sschneider-ihre-pvs** [comment]
> 

doesn't seem to be the case for insiders?

**Comment by sschneider-ihre-pvs** [comment]
might be hidden in the side panel even when enabled :D

**Comment by equiman** [comment]
Today I found a bug trying to install the same extension on different profiles.

If you start to install the extension but don't wait to finish and switch to other profiles, the extension never changes the message again and freezes the button to 'installing' (need VScode reload)

**Comment by sandy081** [comment]
@equiman This is already tracked here - https://github.com/microsoft/vscode/issues/162645

**Comment by moseleyi** [comment]
> > Will we be able to assign a profile to language/folder/workspace?
> 
> The way I see it you could already assign settings to a language even before profiles (e.g. the `"[<language>]"` syntax in `settings.json`) whereas the idea of profiles as I get it is for them to be easily switchable. I guess you can expand on that idea by using language-specific settings inside a profile's `settings.json`. As for folders and workspaces, some previous comments mention that this is already working like that.

Really? Can I set different indentation based on folder+language in settings.json? It's only to conform to rules set by my clients

**Comment by studgeek** [comment]
Excited for this feature! Is there a single place where the current documentation for Settings Profiles is located, or do we have to read through the last few release notes?

And what is the precedence of a profile relative to user, workspace, and folder? They seem to always be stored at user level, but I do think it would be helpful to have profiles that work at the workspace level (and override workspace-level settings). As an example, a user may want interact with a folder/workspace in different ways (e.g. different toolchains or different build targets (which might result in different clang settings, search paths, etc); different focuses between triaging/debugging/coding).

**Comment by mrwensveen** [comment]
@studgeek You could solve the workspace-level settings by using different workspace files (e.g. triage.code-workspace, develop.code-workspace, etc.).

I'm not sure if the profile is between the user and workspace level or if it replaces the user level settings.

**Comment by sandy081** [comment]
@studgeek 

> Excited for this feature! Is there a single place where the current documentation for Settings Profiles is located, or do we have to read through the last few release notes?

This feature is still in preview. We will have a wiki document once we make it stable. 

Profiles are defined at user level and you can associate a folder/workspace or multiple to a profile. They are like an environment in which you want to work. Nevertheless, folder/workspace settings always take precedence because they are defined relatively specific to that workspace/folder. If you think a workspace can be interacted in different environments, I think then they such environment configurations shall not be defined at workspace level. Its VS Code phenomena that if something is defined at workspace level then the workspace shall respect that in any environment.

**Comment by thistlillo** [comment]
How is a `profile` defined? I read this in the upgrade log, but cannot understand what it is exactly:
```
Preview features
Profiles
We have been working the last couple of months to support Profiles in VS Code, which is one of the most popular asks from the community. This feature is available for preview via the workbench.experimental.settingsProfiles.enabled setting. Try it out and give us your feedback by creating issues in the [vscode repository](https://github.com/microsoft/vscode/issues) or commenting in [issue #116740](https://github.com/microsoft/vscode/issues/116740).
```

**Comment by sschneider-ihre-pvs** [comment]
https://code.visualstudio.com/updates/v1_74#_preview-features

**Comment by DiegoBM** [comment]
When you switch profiles, are extensions effectively unloaded from memory? I've two profiles, the default one with barely any extensions or configuration, and another one with a decent amount of extensions, some of them theoretically heavy like the C++ and C# extensions. But when switching from the "heavy" one to the default one, there is barely a difference in memory consumption according to the Task Manager in Windows 11.

**Comment by sandy081** [comment]
Yes, switching profiles shall unload extensions from memory. Can you please take a memory profile and see which process is consuming the memory

**Comment by DiegoBM** [comment]
I can confirm in the "Running Extensions" window that when I switch profiles, there is a small variation in the extensions displayed. Some extensions like the "React Native Tools" extension show right away in that list, whereas  some extensions like "C++" or "C#" won't show. Maybe they won't run until they are used.. I can see though that their related commands appear on the command palette upon profile change, so they must be working normally.

**Comment by DiegoBM** [comment]
I've noticed that when certain Extensions that are disabled for the active profile update (in my case all the "C++" extensions from Microsoft, as well as the "Remote Explorer", and "Remote - SSH" also from Microsoft, maybe it happens with all?), they get re-inserted in my Extensions list, even if I had them disabled for this particular profile, and will remain there from then onwards. I presume that's not an expected behaviour right?

**Comment by KucharczykL** [comment]
I have created a new profile based on the default one and when I try to export the settings, it shows the `Chose (sic) what to export` window that loads endlessly. If I switch back the default profile, exporting it works just fine.

**Comment by sandy081** [comment]
@depascalis Inherit/extending from default profiles is not yet supported, please see this - https://github.com/microsoft/vscode/issues/156144. We will look into this in next phase based on user feedback and requirements.

**Comment by grymmjack** [comment]
Before I knew about this feature I was using the --user-data-dir and --extensions-dir along with this shell script:

```bash
#!/usr/bin/env bash
VSC_PROFILE_ROOT="${HOME}/vscode-profiles"
PROFILE_DIR="${VSC_PROFILE_ROOT}/${1}"
USERDATA_DIR="${PROFILE_DIR}/user-data"
EXTENSIONS_DIR="${PROFILE_DIR}/extensions"

if [ -z "${1}" ]; then
    echo -e "\nUsage: ${0} [profile-name] {workspace.vscode-workspace}"
    echo "Where profile-name is directory storage for vscode profile."
		echo "Where workspace.vscode-workspace is the vscode workspace file."
    echo -e "\nExample: ${0} python"
    echo "Creates ${USERDATA_DIR}, ${EXTENSIONS_DIR}"
    echo -e "Then executes 'code' with --extensions-dir --user-data-dir args in cwd\n"
else
    if [ ! -d "${PROFILE_DIR}" ]; then
        mkdir -p "${PROFILE_DIR}" "${USERDATA_DIR}" "${EXTENSIONS_DIR}"
    fi
		if [ ! -z "${2}" ]; then
    		code --extensions-dir "${EXTENSIONS_DIR}" --user-data-dir "${USERDATA_DIR}" "${2}"
				echo "${0} launching code: --extensions-dir ${EXTENSIONS_DIR} --user-data-dir ${USERDATA_DIR} ${2}"
		else
				code --extensions-dir "${EXTENSIONS_DIR}" --user-data-dir "${USERDATA_DIR}" .
				echo "${0} launching code: --extensions-dir ${EXTENSIONS_DIR} --user-data-dir ${USERDATA_DIR} $(pwd)"
		fi
fi
```

It isn't syncable but it does the job. I mainly use profiles for different languages. I'm really really really looking forward to this profile native support in vscode, esp. since community can build and share entire profiles to get a full-on language support with extensions, settings, etc. This is HUGE. I hate having shared extensions now, and have been using workspaces and workspace settings in my profiles and it has been dreamy.

Thanks devs! <3

**Comment by rubensa** [comment]
To me, one of the most interesting things about profiles is for using it with **Dev Containers** to allow extension customization as per developer but looks like this is not supported at the time as I'm getting:

```[error] Installing extensions to a specific profile is not supported in remote scenario: Error: Installing extensions to a specific profile is not supported in remote scenario```

So I can't install extra extensions, export the profile, rebuild the container and import the profile configuration so my custom extensions are added to the extensions installed by the **devcontainer.json** extensions tag.

@miguelsolorio  Are there any plans to support this scenario in the near future?

**Comment by sandy081** [comment]
@rubensa Not sure if I understand your use case here to support profiles in Dev Containers. I think Dev Container itself is a profile. Can you please elaborate more here. FYI, we have a feature request to support profiles for remote - https://github.com/microsoft/vscode/issues/165247. Please comment and provide your scenario there.

**Comment by DiegoBM** [comment]
I've noticed that if you move one project from one folder to a different one then vscode won't remember the path anymore, which makes me assume that it's vscode the one storing the map between maps and paths for itself. Wouldn't it be more practical to store the profile information withing the project's .vscode folder? Or at least to allow to override the global setting from there?

**Comment by rubensa** [comment]
@sandy081 By **Dev Containers** I mean [this](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extension that allows you to develop inside a **Docker** container.
With that extension, in the **devcontainer.json** file, you can specify the extensions that are going to be installed on **Docker** container first initialization.
I'd like to add more extensions afterwards, export the profile, rebuild the container (this makes you loose any previously added extensions) and then re-import the profile (so the extras extensions are installed again).
This should allow to have a predefined set of extensions for the project in the **devcontainer.json** file but add custom extensions as per individual developer (each developer could configure a personal profile).

**Comment by bwateratmsft** [comment]
@isidorn and I were chatting in a meeting some and I had a suggestion that could make profiles even more powerful. In short, if profiles could be _layered_--in a way similar to the layering of default / user / workspace / language / etc. settings--this could provide a lot of flexibility.

As an example, suppose the Python and Azure extensions each release some profiles containing an interesting mix of settings, extensions, snippets, etc. A user wants to use _both_ of these, but if two profiles can't be active at once, or layered in some way, then they cannot.

Later on, suppose they no longer want/need the Python profile, because their next project is in a different language. They may want to turn that Python profile "off" (or remove it), without disturbing the other profile(s) present. Layering of the profiles could help in this scenario as well.

**Comment by Lamarcke** [comment]
> @sandy081 By **Dev Containers** I mean [this](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) extension that allows you to develop inside a **Docker** container. With that extension, in the **devcontainer.json** file, you can specify the extensions that are going to be installed on **Docker** container first initialization. I'd like to add more extensions afterwards, export the profile, rebuild the container (this makes you loose any previously added extensions) and then re-import the profile (so the extras extensions are installed again). This should allow to have a predefined set of extensions for the project in the **devcontainer.json** file but add custom extensions as per individual developer (each developer could configure a personal profile).

I believe this should be filled against the DevContainer repo. This is something which i think should be implemented too. 

My opinion is, since DevContainer extensions are handled somewhat different from all other remote extensions (in the sense that you have a file which specify what extensions are available), you may end with overlapping functionality.

I personally think that a simple "this machine" extension array in the `devcontainer.json` file, or even a separate file that allows you to specify extensions to be enabled (and not only recommended) in the current workspace would solve this issue.

**Comment by sanmai-NL** [comment]
@bwateratmsft See #15909

**Comment by sandy081** [comment]
@rubensa Your requirement is this feature request - https://github.com/microsoft/vscode/issues/165247

**Comment by Issam-seghir** [comment]
if I create a new profile based on the default profile where this profile was created in my windows explorer? 
I do a backup for my vscode folder every period with settings and extensions, I need to know where I can find the folder of the new profile!!

**Comment by sandy081** [comment]
it is stored in the `profiles` folder next to your default profile `settings.json` file.

**Comment by Issam-seghir** [comment]
> it is stored in the `profiles` folder next to your default profile `settings.json` file.

cool, I find my profile settings/snippets now but what about extensions? 
in windows is stored in `C:\Users\User-name\.vscode\extensions`
I can see only one folder of extensions ( for the default profile )

**Comment by sandy081** [comment]
All extensions are installed(downloaded) in the installation folder, and each profile has an extensions manifest file (`extensions.json`) that contains what extensions are included in the profile.

**Comment by jd-solanki** [comment]
Is it possible to enable a few extensions in certain user profile? I don't want to install them on profile switch.

**Comment by Lamarcke** [comment]
> Is it possible to enable a few extensions in certain user profile? I don't want to install them on profile switch.

Can you please expand on what you want to do? If i understand correctly, you just need to uninstall a extension while in any profile to have it disabled for that profile.

**Comment by jd-solanki** [comment]
@Lamarcke Here's my workflow: https://github.com/microsoft/vscode/issues/128030

Globally I have disabled language specific extensions and want to enable them only if I am in that language's workspace/profile. In summary, If I am working on vue I don't want to enable an extension for python.

Do note that I don't want to uninstall them, just enable those I want for that profile.

This will save the resources of the system. For install/uninstall I think it will download and uninstall on each profile switching that is not efficient compared to enable/disable.

**Comment by Lamarcke** [comment]
> @Lamarcke Here's my workflow: https://github.com/microsoft/vscode/issues/128030
> 
> Globally I have disabled language specific extensions and want to enable them only if I am in that language's workspace/profile. In summary, If I am working on vue I don't want to enable an extension for python.
> 
> Do note that I don't want to uninstall them, just enable those I want for that profile.
> 
> This will save the resources of the system. For install/uninstall I think it will download and uninstall on each profile switching that is not efficient compared to enable/disable.

I understand what you are trying to do. Unfortunately, i don't think there's a way to enable/disable extensions in a workspace-basis, at least for now.

However, VS Code doesn't "download and install" your extensions when you switch profiles. Rather, it keeps all your extensions in your user dir, and then just enable the ones you current profile has "installed".

e.g.: if you have a Python and a Vue profile, with the given language extensions installed, the "Vue" plugin gets **disabled** and the Python plugin gets **enabled** when you change to the Python profile. This process doesn't involve installation or uninstallations.

That's why switching profiles with different sets of extension is usually really fast.

**Comment by jd-solanki** [comment]
Hi, thanks for giving the insights. Are there docs on user profiles? How can I tell the profile to enable the specific extension?

Out of curiosity, What happens if we don't use profiles feature and we uninstall and install the extension? Does it get downloaded or something else?

Thanks.

**Comment by Lamarcke** [comment]
> Hi, thanks for giving the insights. Are there docs on user profiles? How can I tell the profile to enable the specific extension?
> 
> Out of curiosity, What happens if we don't use profiles feature and we uninstall and install the extension? Does it get downloaded or something else?
> 
> Thanks.

I don't think there's specific docs about Settings Profiles for now, but you can do a quick search for "Profiles" on your browser in updates since [Jun/2022](https://code.visualstudio.com/updates/v1_69#_settings-profiles). 

What i do **think** happens is that only when no Profile (the "default" mode is also considered a profile) is using a extension, the extension files are truly removed. While i did say that changing profiles doesn't involve the usual install-uninstall process, actually clicking "Install" or "Uninstall" on extensions does. 

Back to my example, if you are on the Python profile, clicking on "Uninstall" on the "isort" extension (just an example), you are actually uninstalling it, and you will likely need to click the "Reload Window" button or do it manually.

One thing i noticed, is that changing profiles removes the need to "Reload Window" when uninstalling extensions. Again, i **think** VS Code is actually soft-reloading the window for you when you change profiles. I always manually do it though, just in case.

**Comment by Issam-seghir** [comment]
@Lamarcke 
If I understood what you said, then you mean that the two profiles share the same extensions folder, in windows `C:\Users\User-name\.vscode\extensions`
And when, for example, switching to the second profile (non-default one) and uninstall an extension on it, the profile does not delete the extension from the extensions folder, it only disables it on this profile, while if I uninstall it from the default profile, it will be deleted from the extension folder?

**Comment by Lamarcke** [comment]
> @Lamarcke If I understood what you said, then you mean that the two profiles share the same extensions folder, in windows `C:\Users\User-name\.vscode\extensions` And when, for example, switching to the second profile (non-default one) and uninstall an extension on it, the profile does not delete the extension from the extensions folder, it only disables it on this profile, while if I uninstall it from the default profile, it will be deleted from the extension folder?

Please take what i'm saying with a grain of salt. Only contributors can safely confirm this info. This is just from some tests i've done myself.

About your question, i tried testing it myself and this is the result:

I have the "One Dark" theme installed on both the `Default` and `Web` profiles in my VS Code.
The one dark extension folder is named `zhuangtongfa.material-theme-3.15.6` in `C:\Users\User-name\.vscode-insiders\extensions`.

Step 0. Switch to the `Default` profile.
Step 1. Remove `One Dark` from my `Default` profile. This includes using the "Reload Window" command.
Step 2. Check if the files are removed. They are not.
Step 3. Switch to the `Web` profile. There, One Dark is enabled and the switch is instantaneous.
Step 4. Remove `One Dark` from my `Web` profile. 
Step 5. Check if the files are removed. They are not.
Step 6. Close VS Code. Files are still there.
Step 7. Reopen VS Code. Files are gone when VS Code finishes booting up.

I tried this with the inverse order (removing from "Default" and then from "Web" and the behavior is consistent.

This leads me to believe that VS Code scans your profiles' extensions at startup, and if no profile is using a extension, it's removed for good. 

This also confirms the things i have said above: 
A global extension folder; Extensions are only removed when no profile (including the `Default`) is using them; Extensions in the global extension folder get enabled and disabled in a per-profile basis, meaning no install/uninstall on profile changes.

**Comment by sandy081** [comment]
@Lamarcke your insights are correct. Thanks for explaining it to the users.

**Comment by floatingpurr** [comment]
Hello there! Is there a way to tell VSCode to open up a project with a specific profile? Maybe something that you can set in the Workspace settings

**Comment by Lamarcke** [comment]
> Hello there! Is there a way to tell VSCode to open up a project with a specific profile? Maybe something that you can set in the Workspace settings

If i'm not mistaken, VS Code should remember the profile you have choosen for a given workspace (if you select the profile while the workspace is open).

Can someone please confirm?

**Comment by andywillis** [comment]
It looks like it does @Lamarcke @floatingpurr but where is it stored? Is it only stored locally, or is a reference to the workspace/folder stored with the profile which is then synced? It might be useful if that information could be explicitly set in the workspace settings too for reference.

**Comment by Lamarcke** [comment]
> It looks like it does @Lamarcke @floatingpurr but where is it stored? Is it only stored locally, or is a reference to the workspace/folder stored with the profile which is then synced? It might be useful if that information could be explicitly set in the workspace settings too for reference.

I believe it's only stored locally. In my opinion, it wouldn't make sense to have it defined in a workspace file because there's no way to define profiles in a per-workspace basis. It would also be redundant if [this](https://github.com/microsoft/vscode/issues/40239) is ever implemented.

**Comment by LasseRosenow** [comment]
Is inheritance of the profiles still planned?

I tried out settings profiles and created some, but they seem not to follow changes in my default profile.

This kinda makes it a bit useless for me, because I do like to for example enable experimental features some times, or change the font etc. and it would be a shame if I have to do this for each and every profile seperately.

**Comment by sandy081** [comment]
@LasseRosenow inheriting from default profile is tracked here - https://github.com/microsoft/vscode/issues/156144. It would be great if you can provide your usecase in that issue.

**Comment by zjffun** [comment]
How `ExtensionContext.globalStorageUri` work with profiles? We use `vscode.Uri.joinPath(context.globalStorageUri, "../../snippets")` to get user snippets, and find `globalStorageUri` will not change with profiles. https://github.com/zjffun/vscode-snippets-manager/issues/8

VS Code uses `userDataProfileService.currentProfile.snippetsHome` internally, could `snippetsHome` provide to extensions?

https://github.com/microsoft/vscode/blob/97dec172d3256f8ca4bfb2143f3f76b503ca0534/src/vs/workbench/contrib/snippets/browser/commands/configureSnippets.ts#L240-L244

Thank you.

**Comment by sandy081** [comment]
`ExtensionContext.globalStorageUri` always points to same location irrespective of profiles. Please follow up in this issue - https://github.com/microsoft/vscode/issues/160466

**Comment by xxxxj-up** [comment]
I read the issues. The ideas are very interesting, but I hope they are not limited to modifying the global setting file manually, but like Windows, Only simple operations are needed. For example, extension sets can be moved like files. Extensions can be placed in different extension sets. At the same time, simple operations can be used to set what extension sets and setting files the environment needs. The whole process does not need to display and modify the global setting file. At this point, the issue can be closed

**Comment by sandy081** [comment]
Closing this plan item as this feature is Generally Available from next stable. Please file separate issues for any other requirements or bugs. Thanks every one of the feedback.

**Comment by Lamarcke** [comment]
> Closing this plan item as this feature is Generally Available from next stable. Please file separate issues for any other requirements or bugs. Thanks every one of the feedback.

I'm kinda new to VS Code Insiders. Is it already available in Insiders? Thank you for the great work.

**Comment by xxxxj-up** [comment]
> 关闭此计划项，因为此功能已从下一个稳定版正式发布。请针对任何其他要求或错误提交单独的问题。感谢每一个反馈。

Can you tell me what new content has been added to this problem

**Comment by sandy081** [comment]
> I'm kinda new to VS Code Insiders. Is it already available in Insiders? Thank you for the great work.

Yes, and will be available in next stable release too.

**Comment by anibalsanchez** [comment]
Thanks for **Profiles**. It is the best feature ever.

I've just started using it with eight profiles, including one "Airplane Mode" [😎](https://apps.timwhitlock.info/emoji/tables/unicode#emoji-modal)

**Comment by equiman** [comment]
Where can I find the profiles documentation?

**Comment by sandy081** [comment]
> I was also surprised that they released a new feature without actually telling us how to use it.

Sorry about that, and we will be coming up with user guide soon. I thought that the summary in release notes - https://code.visualstudio.com/updates/v1_75#_profiles is good enough to start with. But I agree user guide is definitely needed.

**Comment by anonymous1184** [comment]
@sandy081 Any command line to start a specific profile?

**Comment by gjsjohnmurray** [comment]
> @sandy081 Any command line to start a specific profile?

Please use `code --help` to learn what arguments you can specify.

**Comment by equiman** [comment]
> @sandy081 Any command line to start a specific profile?

Yes, you can use the `--profile` flag.

code --profile "Default"
code --profile "DotNet"
code --profile "React"
code --profile "Node"

I've also defined some aliases that help open the current path using some profiles.

```
alias vc="code --profile \"Default\""
alias vco="vc ."
alias vcn="vc . -n"
alias vd="code --profile \"DotNet\""
alias vdo="vd ."
alias vr="code --profile \"React\""
alias vro="vr ."
alias vn="code --profile \"Node\""
alias vno="vn ."
```

**Comment by andywillis** [comment]
Loving the new profiles feature. However, at the moment it's not very clear _what_ profile is being used. Ideally a project/workspace would trigger a specific profile to load but, failing that, an indicator on the status bar would be hugely useful.
